julia/core.py

```diff
@@ -277,24 +287,61 @@ def juliainfo(runtime='julia'):
          if PyCall_depsfile !== nothing && isfile(PyCall_depsfile)
              include(PyCall_depsfile)
              println(pyprogramname)
+             println(libpython)
          end
-         """])
+         """],
+        # Use the original environment variables to avoid a cryptic
+        # error "fake-julia/../lib/julia/sys.so: cannot open shared
+        # object file: No such file or directory":
+        env=_enviorn)
     args = output.decode("utf-8").rstrip().split("\n")
-    if len(args) == 3:
-        args.append(None)  # no pyprogramname set
+    args.extend([None] * (len(JuliaInfo._fields) - len(args)))
     return JuliaInfo(*args)
 
 
 def is_same_path(a, b):
-    a = os.path.normpath(os.path.normcase(a))
-    b = os.path.normpath(os.path.normcase(b))
+    a = os.path.realpath(os.path.normcase(a))
+    b = os.path.realpath(os.path.normcase(b))
     return a == b
 
 
-def is_different_exe(pyprogramname, sys_executable):
-    if pyprogramname is None:
+def is_compatible_exe(jlinfo, _debug=lambda *_: None):
+    """
+    Determine if Python used by PyCall.jl is compatible with this Python.
+
+    Current Python executable is considered compatible if it is dynamically
+    linked to libpython (usually the case in macOS and Windows) and
+    both of them are using identical libpython.  If this function returns
+    `True`, PyJulia use the same precompilation cache of PyCall.jl used by
+    Julia itself.
+
+    Parameters
+    ----------
+    jlinfo : JuliaInfo
+        A `JuliaInfo` object returned by `juliainfo` function.
+    """
+    _debug("jlinfo.libpython =", jlinfo.libpython)
+    if jlinfo.libpython is None:
+        _debug("libpython cannot be read from PyCall/deps/deps.jl")
+        return False
+
+    if determine_if_statically_linked():
+        _debug(sys.executable, "is statically linked.")
+        return False
+
+    # Note that the following check is OK since statically linked case
+    # is already excluded.
+    if is_same_path(jlinfo.pyprogramname, sys.executable):
+        # In macOS and Windows, find_libpython does not work as good
+        # as in Linux.  We add this shortcut so that PyJulia can work
+        # in those environments.
```

Comment:
I needed this shortcut to make it work since `find_libpython.py` and `PyCall/deps/build.jl` sometimes disagree:


For example, in macOS, `PyCall/deps/build.jl` finds:

> [L957](https://travis-ci.org/JuliaPy/pyjulia/jobs/421978787#L957) `jlinfo.libpython = /usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/lib/libpython3.7m`

but `find_libpython.py` finds:

> [L958](https://travis-ci.org/JuliaPy/pyjulia/jobs/421978787#L958) `py_libpython = /usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/Python`


Another example in Windows: `PyCall/deps/build.jl` cannot find a fullpath:

> [L517](https://ci.appveyor.com/project/Keno/pyjulia/build/1.0.174/job/gj7k6da28y0wq3rd#L517) `jlinfo.libpython = python27`

but `find_libpython.py` "finds" it:

> [L518](https://ci.appveyor.com/project/Keno/pyjulia/build/1.0.174/job/gj7k6da28y0wq3rd#L518) `py_libpython = C:\Windows\system32\python27.dl`

(But it looks like the same path was found in win32 and win64.  Ultimately, we need someone to write a function that uses `dladdr` equivalent in Windows.  It should be possible by just extending: https://stackoverflow.com/a/16659821)


I think it's better to use the same script to discover libpython in PyCall.jl and PyJulia.  We can then avoid a hack like this.  Since `find_libpython.py` can call `libdl.dladdr` (though it works only if libpython is dynamically linked), I think it's better approach.

Reply:
In the cases where they disagree, which is correct?  Should deps/build.jl be fixed?

Reply:
I don't have access to macOS so I'm not sure. But PyJulia printed

```
jlinfo.libpython = /usr/local/opt/python/Frameworks/Python.framework/Versions/3.7/lib/libpython3.7m
py_libpython = /usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/Python
jl_libpython = None
```

where `jlinfo.libpython` is `libpython` saved in `deps/deps.jl` and `jl_libpython` is the path I tried to resolved by

https://github.com/JuliaPy/pyjulia/blob/6982ae885989285fc58c4b5181071b79975ea7f7/julia/find_libpython.py#L176-L182

So `jl_libpython = None` means that `libpython3.7m.dylib` did not exist (I set `suffix = ".dylib"` in macOS).  Maybe we should try `libpython3.7m.so` in macOS too?

In the case of Windows, `deps/build.jl` just gave up (`jlinfo.libpython = python27`) whereas `ctypes.util.find_library` finds `C:\Windows\system32\python27.dl`.

Reply:
I'm trying it: 1d599be8d1d2e3dbf68b18ce07afc1afeea6ed51